messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java

```diff
@@ -148,6 +185,29 @@ protected void decode(
     }
   }
 
+  static String readString(ByteBuf buffer, int length, Charset charset) {
+    if (buffer.isDirect()) {
+      final String result = buffer.toString(buffer.readerIndex(), length, charset);
+      buffer.skipBytes(length);
+      return result;
+    } else {
+      if (buffer.hasArray()) {
+        final int offset = buffer.arrayOffset() + buffer.readerIndex();
+        final String result;
+        if (offset == 0 && length == buffer.array().length) {
+          result = new String(buffer.array(), charset);
+        } else {
+          result = new String(buffer.array(), offset, length, charset);
+        }
+        buffer.skipBytes(length);
+        return result;
+      }
+      final byte[] array = new byte[length];
```

Comment:
Complicated branches. Why not if(isDirect) elseif (hasArray) else?

Reply:
sgtm, updated.